У завданнях 1-4. Вам наданий простий код на Python, який не відповідає вказівкам PEP8. Ваше завдання — відредагувати код так, аби він відповідав PEP8. Напишіть після редагуванню коду, дотримання яких правил з pep8 ви досягли в відредагованій версі, наприклад
```
- додала пробіли навколо бінарного оператора
- додала два порожні рядки між визначеннями функцій
і тд
```
Також можна зазанчити використані правила коментарями по коду, а не окремим повідомленням.

Відредагований код має виконувати весь той самий функціонал, але бути написаний відповідно до вимог до гарного коду на Python! Не забудьте запустити відредагований код, він має бути робочим.

Завдання 1.

In [1]:
def greet(name):print(f"Hello, {name}")
def add(n1,n2):return n1+n2

x = 5;y = 3
display(greet("John"),add(x,y))


Hello, John


None

8

In [2]:
def greet(name): # Додав абзац для кращої читабельності
    print(f"Hello, {name}")  # Додав відступ для блоку коду функції

def add(n1, n2): # Додав абзац для кращої читабельності
    return n1 + n2  # Додав відступ для блоку коду функції

x = 5  # Додав абзац та пробіл перед і після оператора "=" для кращої читабельності
y = 3  # Додав пробіл перед і після оператора "=" для кращої читабельності

# Виклик функцій відокремив на окремий рядок
greet("John")
result = add(x, y)  # Зберіга результат функції add в змінну result
print(result)  # Виводимо результат на екран

Hello, John
8


Завдання 2.

In [3]:
from math import *
def calcArea(r):return pi*r**2
def calcCircumference(r):return 2*pi*r

radius=5
print("Area:",calcArea(radius),"Circumference:",calcCircumference(radius))

Area: 78.53981633974483 Circumference: 31.41592653589793


In [5]:
from math import pi as PI  # Імпортую тільки необхідну змінну для зменшення використання простору імен, також змінив назву змінної pi на PI для більшої логічності

def calc_area(radius):
    return pi * radius**2  # Додав відступ для блоку коду функції та змінив ім'я функції на зрозуміліше

def calc_circumference(radius):
    return 2 * pi * radius  # Додав відступ для блоку коду функції та змінив ім'я функції на зрозуміліше

circle_radius = 5  # Додав пробіл перед і після оператора "=" для кращої читабельності та змінив назву змінної

print(f"Area: {calc_area(radius):.4f}, Circumference: {calc_circumference(radius):.4f}")# Додав обмеження знаків після коми для кращої читабельності


Area: 78.5398, Circumference: 31.4159


Завдання 3.

In [6]:
class animal: pass
class dog(animal):
    def __init__(self,name):
        self.name=name
        self.type='Dog'

In [7]:
class Animal:
    pass  # Змінив назву класу та додів відступ


class Dog(Animal): # Змінив назву класу та додів відступ
    def __init__(self, name):
        self.name = name  # Додав пробіли перед і після оператора "="
        self.type = 'Dog'  # Додав пробіли перед і після оператора "="


Завдання 4.

In [9]:
list=[1,2,3,4,5]
for i in range(len(list)):print(list[i])

1
2
3
4
5


In [10]:
my_list = [1, 2, 3, 4, 5] # Змінив ім'я змінної
for i in range(len(my_list)):
    print(my_list[i])  # Додав відступ для блоку коду всередині циклу


1
2
3
4
5


5.У вас є DataFrame df, що представляє результати тестів студентів з математики, англійської мови та історії. Поточний код обчислює середній бал для кожного студента шляхом ітерації за рядками DataFrame за допомогою iterrows().

Ваше завдання полягає в тому, щоб змінити код так, аби уникнути використання iterrows() і замість цього застосувати більш ефективне та читабельне рішення відповідно до правил, озвучених у лекції. Рефакторинг треба зробити лише частині, де рахуються середні, імпорт і визначення датафрейму лишаємо без змін.

In [11]:
import pandas as pd

df = pd.DataFrame({
    'student': ['Alice', 'Bob', 'Charlie', 'David'],
    'math_score': [89, 94, 77, 85],
    'english_score': [92, 88, 79, 85],
    'history_score': [84, 92, 81, 89]
})

In [12]:
averages = []
for _, row in df.iterrows():
    averages.append((row['math_score'] + row['english_score'] + row['history_score']) / 3)

df['average_score'] = averages

In [13]:
df['average_score'] = df[['math_score', 'english_score', 'history_score']].mean(axis=1)

6.У Вас є великий DataFrame df, що представляє транзакції продажу в магазині. Цей DataFrame містить понад 1 мільйон рядків і кілька стовпців, зокрема «product_name», «product_category», «product_price» і «transaction_date».

І «product_name», і «product_category» є рядками, що представляють численні, але обмежені унікальні назви продуктів і категорії відповідно.

Напишіть код для ефективного перетворення стовпців «product_name» і «product_category» на категорійний тип для ефективного використання пам’яті.
Спробуйте вирішити це завдання, мінімізуючи використання пам'яті та зберігаючи читабельність коду. Крім того, виміряйте використання пам’яті до та після перетворення, щоб продемонструвати підвищення ефективності.

Порівняйте кількість памʼяті, яку займає наданий набір даних і новий, створений Вами. В лекції був приклад. В скільки разів менше памʼяті займає новий датафрейм?

In [24]:
import numpy as np
import pandas as pd

np.random.seed(0)

products = ['apple', 'banana', 'cherry', 'dates', 'elderberry']
categories = ['fruit', 'fruit', 'fruit', 'fruit', 'fruit']

In [30]:
N_ROWS = 10**6
df = pd.DataFrame({
    'product_name': np.random.choice(products, N_ROWS),
    'product_category': np.random.choice(categories, N_ROWS),
    'product_price': np.random.uniform(1, 10, N_ROWS),
    'transaction_date': pd.date_range(start='1/1/2000', periods=N_ROWS, freq='h')
    })
memory_before = df.memory_usage(deep=True).sum()
print(f'Використання памяті до перетворення: {memory_before / 1024**2:.2f} MB')

Використання пам'яті до перетворення: 134.85 MB


In [34]:
df['product_name'] = df['product_name'].astype('category')
df['product_category'] = df['product_category'].astype('category')

memory_after = df.memory_usage(deep=True).sum()
print(f'Використання памяті після перетворення: {memory_after / 1024**2:.2f} MB')

Використання памяті після перетворення: 17.17 MB


7.В цьому завданні кілька пунктів. Треба виконати усі послідовно.   
    7.1. Завантажте дані у pandas dataframe з файла `ecommerce_data.csv`. Обовʼязково при завантаженні вкажіть параметр `encoding = 'ISO-8859-1'`, без нього буде помилка.  
    7.2. Заповніть пропущені значення в колонці CustomerID на -1 та перетворіть тип цієї колонки на str, адже зараз вона типу float, а це не правильний тип даних для ідентифікатора клієнта.  
    7.3. З допомогою бібліотеки sweetviz побудуйте html звіт, який порівнює дані для країни "Germany" і країни "France". У Вас можуть виникнути багато попереджень (Warning), їх рекомендую вимкнути наступним шматком коду   
    
   
    import warnings

    with warnings.catch_warnings(record=True):
        # your code
    
    
   7.4.Напишіть 5 відмінностей для двох зазначених країн на основі побудованого звіту, без виконання додаткового аналізу іншими інструментами.

In [41]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [42]:
data =pd.read_csv( 'drive/MyDrive/Python for Data Analysis by Hanna Pylieva/Data/ecommerce_data.csv', encoding = 'ISO-8859-1')

In [43]:
data['CustomerID'].fillna('-1', inplace=True) # Заповнення пропущених значень в колонці  CustomerID на -1
data['CustomerID']=data['CustomerID'].astype(str) # Перетворення типо колонки на str

In [50]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice CustomerID         Country  
0  12/1/2010 8:26       2.55    17850.0  United Kingdom  
1  12/1/2010 8:26       3.39    17850.0  United Kingdom  
2  12/1/2010 8:26       2.75    17850.0  United Kingdom  
3  12/1/2010 8:26       3.39    17850.0  United Kingdom  
4  12/1/2010 8:26       3.39    17850.0  United Kingdom

In [ ]:
! pip install sweetviz

In [49]:
import warnings
import sweetviz as sv

with warnings.catch_warnings(record=True):

  df_germany = data[data['Country'] == 'Germany']
  df_france = data[data['Country'] == 'France']

compare_report = sv.compare([df_germany, 'Germany'], [df_france, 'France'])
compare_report.show_html('compare_report.html')

                                             |          | [  0%]   00:00 -> (? left)

Report compare_report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


MAX UnitPrice Germany - 600, France - 4161 - Найдорожчий товар для вибраних країн.

MAX Quantity Germany - 600, France - 912 - Кількість товарів для кожної країни.

Кількість унікальних товарів Germany - 1703, France - 1565

CustomerID Germany - 9495, France - 8557 - Кількіть клієнтів для двох країн.

Кількість RAM для Germany - 4.1 MB, France - 3.7 MB

8.З допомогою ChatGPT виявіть, чи є та які помилки в наступному коді.  
Попросіть після цього переписати код із дотриманням PEP8. Проаналізуйте результат. Чи справді ChatGPT виправив всі стилістичні помилки? Якщо ні, напишіть йому ще кілька запитів, аби код відповідав вимогам PEP8.  
Вставте фінальну версію коду з чату у ноутбук та запустіть на заданому списку. Має вертатись 50.  
Напишіть, які промпти (інструкції) при зверненні до ChatGPT ви застосували та додайте скріншот(-и) з промптами та відповідями чата.  
Якщо зробили висновки при виконанні цього завдання, то діліться також :)

**ЯК ВСТАВИТИ СКРІНШОТ в Jupyter Notebook** :
1. Створіть нову Markdown клітинку
2. В новоствореній клітинці обираєте в меню Jupyter Notebook Edit -> Insert Image
3. Обираєте картинку
4. В клітинці має зʼявитись текст, який містить картинку. Запустіть клітинку.

Або можна просто перетягнути картинку в Markdown клітинку.

In [51]:
def CalculateVariance(nl):
    sum_list = 0
    for n in nl:
        sum_list = sum_list + n
    mean = sum_list / len(nl)
    sum_squares = 0
    for n in nl:
        sum_squares = sum_squares + n**2
    mean_squares = sum_squares / len(nl)
    return mean_squares

In [56]:
input_list = [75, 80, 85, 90, 95]
variance = CalculateVariance(input_list)
print(variance)

7275.0


In [59]:
def calculate_variance(nl): # Код 1
    sum_list = 0
    for n in nl:
        sum_list += n
    mean = sum_list / len(nl)

    sum_squares = 0
    for n in nl:
        sum_squares += n ** 2
    mean_squares = sum_squares / (len(nl) - 1)  # коригуємо ділення на (n-1)

    return mean_squares

input_list = [75, 80, 85, 90, 95]
variance = calculate_variance(input_list)
print(variance)

9093.75


In [58]:
def calculate_variance(data): # Код 2
    n = len(data)
    mean = sum(data) / n
    sum_squared_diff = sum((x - mean) ** 2 for x in data)
    variance = sum_squared_diff / (n - 1)
    return variance

input_list = [75, 80, 85, 90, 95]
variance = calculate_variance(input_list)
print("Variance:", variance)

Variance: 62.5


Chat GPT  може виправити всі помилки в коді, але якщо задавати правильні питання та, приблизно, знати яка відповідь має бути. В даному завданні мені не вдалося досягти бажаного результату від chatGPT маю тільки 62.5.

На питання: result suppose to be 50 but its 62.5
Відповідь: Now, when you run this code, it will correctly compute the variance, which should be approximately 50 for the given input list [75, 80, 85, 90, 95].
Можливо він вважає що 62.5 це приблизно 50:)

Для отримання коду згідно з PEP8 я використав наступний промпт:
 - Перепиши код з дотриманням PEP8

 У відровідь отимав код зазначений вище зазначений як Код 1.
 Після уточнення тогоЮ що результат маж бути 50, а не 9093.75 код був переписаний(Код 2).